Transform gradients in strokes

adroitwhiz opened the pull request:
### Resolves

Resolves #142

### Proposed Changes

This PR extends the gradient-transforming logic in `transformStrokeWidths` to transform gradients not only in fills (as it previously did) but in strokes as well.

### Reason for Changes

Strokes can also have gradients (and https://github.com/LLK/scratch-paint/pull/1004 makes this much more likely). Previously, they weren't being taken into account when transforms were being applied.

### Test Coverage

Tested manually

adroitwhiz committed:
Transform gradients in strokes

adroitwhiz commented on src/transform-applier.js:
```diff
@@ -505,14 +505,42 @@ const _parseUrl = (value, windowRef) => {
+        } else {
+            try {
+                doc.body.appendChild(svgSpot);
+                const svg = SvgElement.set(windowRef.document.createElementNS(SvgElement.svg, 'svg'));
+                const path = SvgElement.set(windowRef.document.createElementNS(SvgElement.svg, 'path'));
```
```suggestion
                const svg = SvgElement.set(doc.createElementNS(SvgElement.svg, 'svg'));
                const path = SvgElement.set(doc.createElementNS(SvgElement.svg, 'path'));
```
Should I include this code style enhancement as part of this PR?

fsih replied:
Sure

adroitwhiz commented on src/transform-applier.js:
```diff
@@ -505,14 +505,42 @@ const _parseUrl = (value, windowRef) => {
-                fill = element.attributes.fill.value;
+            if (element.attributes) {
+                if (element.attributes.fill) fill = element.attributes.fill.value;
+                if (element.attributes.stroke) stroke = element.attributes.stroke.value;
```
Why are we checking if `element.attributes` exists here but not checking it in the previous if-block (`if (element.attributes['stroke-width'])`)?

fsih replied:
It looks like element.attributes should exist if element.tagName exists, which is what we check for in _isContainerElement, except on Internet Explorer. So the if above would break on IE (but we don't support IE anyway)

adroitwhiz committed:
Use doc instead of windowRef.document

fsih commented on src/transform-applier.js:
```diff
@@ -522,56 +550,42 @@ const transformStrokeWidths = function (svgTag, windowRef, bboxForTesting) {
             element.removeAttribute('stroke-width');
             element.removeAttribute('fill');
+            element.removeAttribute('stroke');
         } else if (_isPathWithTransformAndStroke(element, strokeWidth)) {
```
We should double check if this code is working for stuff besides paths with strokes, like shapes

fsih commented on src/transform-applier.js:
```diff
@@ -568,8 +574,15 @@ const transformStrokeWidths = function (svgTag, windowRef, bboxForTesting) {
+                    const newStrokeRef = _createGradient(strokeGradientId, svgTag, bbox, matrix);
+                    if (newStrokeRef) stroke = newStrokeRef;
+                }
             }
```
It seems like fillGradientId and strokeGradientId might be the same, and if they're the same, we should try not to make a duplicate definition (although I'm not sure if it hurts anything if we do have duplicates)

adroitwhiz committed:
Reuse existing transformed gradients

adroitwhiz committed:
Move element categories into own file+correct them

Instead of checking if an element is a graphics element,
check if it's "paintable" (fill/stroke applies to it).

This includes several text-related elements,
and excludes `use` and `image` elements.

fsih reviewed (approved):
Looks good! Let's merge this after the deploy.
